agent_identity/src/document/aggregate.rs

```diff
@@ -78,139 +97,141 @@ impl Aggregate for Document {
-                                        wallet_address.to_string(),
-                                    ));
-                                }
-                                Err(other_error) => return Err(IotaClientError(other_error)),
+                            // Create a new 'blank' DID Document.
+                            IotaDocument::new(&network_name)
+                        });
+
+                        info!("Testing whether a DID Document can be published...");
+
+                        // This code block is doing a dummy publish to ensure that the DID Document is created and can
+                        // be published/updated later. It uses a gas budget of 0 to avoid actually publishing the
+                        // document.
+                        let document = match identity_client.publish_did_document_update(document.clone(), 0).await {
+                            // This match arm can never be reached, because we use a gas budget of 0.
+                            Ok(document) => document,
+                            // This error occurs when the DID Document is not published yet. We will not return an
+                            // error because the `PublishDocument` command will handle the actual publishing if the
+                            // funds are sufficient.
+                            Err(identity_iota::iota::rebased::Error::DIDResolutionError(_err)) => {
+                                warn!("Document is not published yet.");
+
+                                document
                             }
-                        }
+                            // This error occurs when the `identity_client` has no control over the DID Document that
+                            // has been stored in the aggregate instance. This usually means that the current keys
+                            // stored in the KMS have been updated between boots. We throw an error here indicating
+                            // that the original KMS needs to be used or that the database needs to be wiped.
+                            // TODO: implement KMS migration.
+                            Err(identity_iota::iota::rebased::Error::Identity(err))
+                                if err.contains("address") && err.contains("has no control over Identity") =>
+                            {
+                                warn!("Identity control lost: {err}");
+
+                                return Err(DocumentError::IotaControllerError(
+                                    identity_iota::iota::rebased::Error::Identity(err),
+                                ));
+                            }
+                            // This error is to be expected because we use a gas budget of 0.
+                            Err(identity_iota::iota::rebased::Error::TransactionUnexpectedResponse(err))
+                                if err.contains("Gas budget: 0 is lower than min: 1000000") =>
```

Comment:
Isn't this a brittle if-check since the number `1000000` could be changing without our knowledge?

Reply:
yeah you're right, I removed the `1000000` part. In general doing this string matching on the error messages is not ideal (and brittle). Before the IOTA Rebase update identity.rs had better typed errors that we could match on, but now with this update those error types are wrapped inside an external Error type which simply stringifies the error messages.

I had a quick chat with the identity.rs team about this and they've taken it as feedback to try to provide the strong error types again in a future update.